blockchain/chain_sync/src/tipset_syncer.rs

```diff
@@ -0,0 +1,1741 @@
+            parent_tipsets.push(tipset);
+        }
+    }
+    // Unwrapping is safe here because we assume that the tipset
+    // vector was initialized with a tipset that will not be removed
+    let oldest_tipset = parent_tipsets.last().unwrap().clone();
+    // Determine if the local chain was a fork.
+    // If it was, then sync the fork tipset range by iteratively walking back
+    // from the oldest tipset synced until we find a common ancestor
+    if oldest_tipset.parents() != current_head.parents() {
+        const FORK_LENGTH_THRESHOLD: u64 = 900;
+        let fork_tipsets = network
+            .chain_exchange_headers(None, oldest_tipset.parents(), FORK_LENGTH_THRESHOLD)
+            .await
+            .map_err(|err| TipsetRangeSyncerError::NetworkTipsetQueryFailed(err.to_string()))?;
+        let mut potential_common_ancestor =
+            chain_store.tipset_from_keys(current_head.parents()).await?;
+        let fork_length = 1;
+        for (i, tipset) in fork_tipsets.iter().enumerate() {
+            if tipset.epoch() == 0 {
+                return Err(TipsetRangeSyncerError::ForkAtGenesisBlock(format!(
+                    "{:?}",
+                    oldest_tipset.cids()
+                )));
+            }
+            if potential_common_ancestor == *tipset {
+                // TODO: Iterate over the blocks in the forks tipsets,
+                // validate them, and add them to the accepted blocks vector
+                //
+                // Remove elements from the vector since the Drain
+                // iterator is immediately dropped
+                let mut fork_tipsets = fork_tipsets;
+                fork_tipsets.drain((i + 1)..);
+                parent_tipsets.extend_from_slice(&fork_tipsets);
+                break;
+            }
+
+            // If the potential common ancestor has an epoch which
+            // is lower than the current fork tipset under evaluation
+            // move to the next iteration without updated the potential common ancestor
+            if potential_common_ancestor.epoch() < tipset.epoch() {
+                continue;
+            }
+            // Increment the fork length and enfore the fork length check
+            if fork_length > FORK_LENGTH_THRESHOLD {
+                return Err(TipsetRangeSyncerError::ChainForkLengthExceedsMaximum);
+            }
+            // If we have not found a common ancestor by the last iteration, then return an error
+            if i == (fork_tipsets.len() - 1) {
+                return Err(TipsetRangeSyncerError::ChainForkLengthExceedsFinalityThreshold);
+            }
+            potential_common_ancestor = chain_store
+                .tipset_from_keys(potential_common_ancestor.parents())
+                .await?;
+        }
+    }
+    Ok(parent_tipsets)
+}
+
+fn sync_tipset<
+    DB: BlockStore + Sync + Send + 'static,
+    TBeacon: Beacon + Sync + Send + 'static,
+    V: ProofVerifier + Sync + Send + 'static,
```

Comment:
Nit: Probably better to put these in a `where` clause?

Reply:
It's a good question, @creativcoder, but I think if we are going to go into nits that are ambiguous as to whether they add value to the code, we should just maintain a style guide. I don't want to hold up merging this PR over nits.

Reply:
Yeah, me neither. It's a "question" for the very same reason. Jorge can resolve it if it's fine this way.

Reply:
This form of  trait bounds for type parameters on stand alone functions doesn't bother me. Will keep as is for now.